### DISCUSSION TITLE
Readme: shouldn't Direct3D start with 2?

### DISCUSSION THREAD

**Discussion by eddmanx** [elishacloud/dxwrapper#322] [General]
AFAIK DirectX 1 doesn't have D3D; it was first introduced with DirectX 2.

The internal versions are different, but externally D3D is named in relation to the DX name.

**Comment by elishacloud** [comment]
I am not sure how DirectX and Direct3D map to each other.  However, there is most certainly a Direct3D version 1.  Here are the first 4 versions of Direct3D interfaces from Microsoft's public documentation:

* [Direct3D](https://github.com/apitrace/dxsdk/blob/d964b66467aaa734edbc24326da8119f5f063dd3/Include/d3d.h#L203-L221)
* [Direct3D2](https://github.com/apitrace/dxsdk/blob/d964b66467aaa734edbc24326da8119f5f063dd3/Include/d3d.h#L247-L265)
* [Direct3D3](https://github.com/apitrace/dxsdk/blob/d964b66467aaa734edbc24326da8119f5f063dd3/Include/d3d.h#L292-L313)
* [Direct3D7](https://github.com/apitrace/dxsdk/blob/d964b66467aaa734edbc24326da8119f5f063dd3/Include/d3d.h#L346-L363)

You can see a list of all the versions of all interfaces from [here](https://github.com/elishacloud/dxwrapper/tree/master/ddraw/Versions).

Furthermore, I know games use these v1 interfaces.  For example, [Moto Racer](https://www.gog.com/en/game/moto_racer) uses both DirectDraw v1 and Direct3D v1.  There are other games as well, such as [Star Trek Starfleet Academy](https://www.gog.com/en/game/star_trek_starfleet_academy) and [Star Wars X-Wing Alliance](https://www.gog.com/en/game/star_wars_xwing_alliance) that use both DirectDraw v1 and Direct3D v1.

**Reply by elishacloud** [reply]
I am looking at the documentation for each version of DirectX and it appears that:

* DirectX1 uses DirectDraw v1
* DirectX2 uses DirectDraw v1 and Direct3D v1
* DirectX3 uses DirectDraw v2 and Direct3D v1
* DirectX5 uses DirectDraw v3 and Direct3D v2
* DirectX6 uses DirectDraw v4 and Direct3D v3
* DirectX7 uses DirectDraw v7 and Direct3D v7

_Note: I don't have any documentation for DirectX4._

**Reply by eddmanx** [reply]
That's the thing, those are internal and I don't recall publications, or MS (outside of API calls-level dev material), ever referring to them that way. The public names were tied to the DX number to avoid confusion. It was only with DX7 that the internal and external numbers became the same.

There aren't many examples online from back then, but there was an MS official press release referring to DX6's d3d as "Direct3D 6", but google can't find it now for some reason.
This page refers to it as such though: https://www.gamedev.net/articles/programming/graphics/direct3d-7-immediate-mode-framework-programming-r1028/

I think I've seen "Direct3D 5" in game readme files. I remember finding old websites referring to it as such but again google just doesn't find them now.

DirectX 4 doesn't exist; it was canceled.

It might have been that DX2 and probably DX3's d3d were simply called "Direct3D" back then with no number, but retroactively they are called "Direct3D 2" and "Direct3D 3". Feature wise they are apparently the same but IINM 3 has some changes, maybe bug fixes.

**Reply by elishacloud** [reply]
I just posted all the legacy [DirectX documentation](https://github.com/elishacloud/Legacy-DirectX-Documentation) because I could not find them anywhere on the Internet.

> there was an MS official press release referring to DX6's d3d as "Direct3D 6"

Microsoft refers to it as "Direct3D in DirectX 6" in the press release before they ever refer to it as "Direct3D 6".  The references to "Direct3D 6" are clearly just shorthand for "Direct3D in DirectX 6".  It isn't saying they are using Direct3D 6, as that would directly contradict their official documentation.

I searched through the [DirectX6 documentation](https://github.com/elishacloud/Legacy-DirectX-Documentation/tree/main/directx6) and I cannot find a single reference to "Direct3D 6" or "Direct3D6".  However, I can find lots of references to "Direct3D3" and to "DirectX 6".

> I think I've seen "Direct3D 5" in game readme files.

Sounds like marketing to me, just like the press release (which is always marketing).  From a marketing perspective, it sounds better to call it Direct3D 6 rather than Direct3D 3.

I prefer to use the names from the official documentation rather than any marketing materials.

Furthermore, I did a quick Internet search for "Direct3D3" and all of the sites I checked (there were far too many to check them all) were referring to DirectX6, not DirectX3.  [Here](https://github.com/kurlyak/directx61/blob/main/README.md) is an example where is says "DirectX 6.0 Direct3D3 Device".

I already posted the official versions of DirectX and their associated Direct3D versions above.  But I will post them again slightly different format.  These are the official version numbers based on the official Microsoft Documentation:

* DirectX1 = DirectDraw
* DirectX2 = DirectDraw & Direct3D
* DirectX3 = DirectDraw2 & Direct3D
* DirectX5 = DirectDraw3 & Direct3D2
* DirectX6 = DirectDraw4 & Direct3D3
* DirectX7 = DirectDraw7 & Direct3D7

Anyways, to address the title of this discussion, I am using the internal interface versions in my project rather than the DirectX version numbers for two reasons:

1. Some versions of DirectX share the same interfaces and I don't want to duplicate things.
2. I cannot tell the DirectX version many games use.  None of the graphics APIs reveal the DirectX version.  The only thing the game sends is the interface version and games can use mismatching interfaces or older interfaces that would make it unclear which DirectX version it is using.  For example if a game only used DirectDraw, would it be DirectX1 or DirectX2?  With either version of DirectX you can use DirectDraw.  In fact, even in DirectX7 you can chose to just use DirectDraw and not DirectDraw7.  Since there is no way for me to tell the version of DirectX that a game is using I chose to just use the interface versions.

I guess this is the reason starting with DirectX7 Microsoft decided to make the interface versions consistent with the DirectX version to avoid this type of confusion.

**Reply by eddmanx** [reply]
I thought maybe using the non-dev names could be easier for general readers but the interface numbers do work better from a wrapper development standpoint.

The general non-dev materials since 90s have always referred to D3D with the corresponding DX number, well at least since D3D5. I don't think it was exactly for marketing but to not cause confusion for the general non-savvy users. MS' choice of naming certainly wasn't ideal.

The DX/D3D naming system and the way MS connected these two together, as if they are the same thing, caused another headache a while back. On pcgamingwiki people were filling the Direct3D field simply based on the DirectX version that was in the reqs or the DX installer that came with the game. There are games that do use DirectX, but do not use Direct3D at all. For example, a game might really require DirectX 8 and wouldn't work without it, but that's because it uses its sound, controller, etc. part and uses OpenGL for graphics. There were also many similar cases where it does need DX8 to be installed but uses d3dim700.dll or even d3dim.dll.

The interface numbers weren't used though. Many people couldn't tell DX (which is really a package/suite of APIs) and D3D apart, let alone having to deal with a public D3D name and an internal interface name.

D3D8 and 7 games were determined by the dll, and D3D6 and older by tracing dgvoodoo2. Determining between D3D6, 5 and 2/3 was doable. 2 and 3 couldn't exactly be told apart, but they were still listed separately because they don't have the exact same d3dim.dll, and there might be a Direct3D 3 game that maybe won't work properly with the d3dim.dll from the DX2 package.

By the way, regarding the DX documentation availability, they are included with DX SDKs which are all hosted on archive.org. I downloaded them quite a while ago. Comparing the D3D.H files of 2 and 3 showed some slight interface differences but I'm not knowledgeable to know what that difference actually entails.

**Reply by elishacloud** [reply]
> By the way, regarding the DX documentation availability, they are included with DX SDKs which are all hosted on archive.org.

Please let me know where the links are because I couldn't find them.  All I could find was an old SDK for DirectX 7, which doesn't have any documentation for older versions of DirectX.

**Reply by eddmanx** [reply]
I don't know why you couldn't find them. It's very easy 

https://archive.org/search?query=title%3A%28microsoft+directx+sdk%29&sort=title

**Reply by elishacloud** [reply]
> I don't know why you couldn't find them. It's very easy
> 
> https://archive.org/search?query=title%3A%28microsoft+directx+sdk%29&sort=title

Thanks!  I never search archive.org because I always just searched Google and figured that Google can search everything, including archive.org.  But I have not searched in a number of years because I always came up short in the past and figured it was just lost to time, like may other things.
